### PR TITLE
feat: covfie plugin

### PR DESCRIPTION
**Pull request by fredevb**
Added functions for creating fields in covfie from magnetic fields in acts. I.e.:

- Acts::CovfiePlugin::covfieField(interpolatedMagneticField)
- Acts::CovfiePlugin::covfieField(constantBField)

The following method can be used for more a more general conversion by sampling an acts magnetic field provider (which creates a linearly interpolated covfie field):
- Acts::CovfiePlugin::covfieField(magneticFieldProvider, cache, nBins, min, max)

Added tests for conversion of InterpolatedMagneticField, ConstantBField, and SolenoidBField.
Added Python bindings using pybind11 and a simple test.

Furthermore:
- Added --exclude "*build/_deps/*" to clang-tidy (CI)
- Updated cmake options ACTS_BUILD_EVERYTHING to include ACTS_BUILD_PLUGIN_TRACCC (which includes the covfie plugin)

### CODE REVIEW

**Comment by stephenswat** [comment]
Can you explicitly ping me when this is ready for review, please?

**Comment by fredevb** [comment]
@stephenswat it is ready for review :)

File: Examples/Python/src/MagneticField.cpp [hunk 1]
```diff
@@ -45,7 +45,9 @@ void addMagneticField(Context& ctx) {
 
   py::class_<Acts::InterpolatedMagneticField,
              std::shared_ptr<Acts::InterpolatedMagneticField>>(
-      m, "InterpolatedMagneticField");
+      m, "InterpolatedMagneticField")
+      .def("getFieldUnchecked",
+           &Acts::InterpolatedMagneticField::getFieldUnchecked);
```
Comment (stephenswat) [inline Examples/Python/src/MagneticField.cpp:50]: Can you explain what "Unchecked" means here?

File: CMakeLists.txt [hunk 2]
```diff
@@ -157,6 +157,14 @@ set_option_if(ACTS_BUILD_PLUGIN_LEGACY ACTS_BUILD_EVERYTHING)
 set_option_if(ACTS_BUILD_PLUGIN_EXATRKX ACTS_BUILD_EXAMPLES_EXATRKX)
 set_option_if(ACTS_BUILD_PLUGIN_FPEMON
   ACTS_BUILD_EXAMPLES OR ACTS_BUILD_EVERYTHING)
+set_option_if(ACTS_SETUP_ALGEBRAPLUGINS
+  ACTS_BUILD_EXAMPLES OR ACTS_BUILD_EVERYTHING)
+set_option_if(ACTS_SETUP_VECMEM
+  ACTS_BUILD_EXAMPLES OR ACTS_BUILD_EVERYTHING)
```
Comment (stephenswat) [inline CMakeLists.txt:163]: Are these changes related to this PR?

File: Examples/Python/src/MagneticField.cpp (hunk 1 above)
Comment (fredevb) [reply Examples/Python/src/MagneticField.cpp]: I removed it and added python bindings for getField instead (it's used for the python tests).

File: CMakeLists.txt (hunk 2 above)
Comment (fredevb) [reply CMakeLists.txt]: I talked with Paul about the cmake set up. Now setup is enabled by default for vecmem, algebra plugins, covfie, and detray, and the covfie plugin is included in the traccc plugin build option.

**Review by stephenswat — DISMISSED** [review]

File: Plugins/Covfie/src/FieldConversion.cpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  return field;
+}
+
+/// @brief Generate the affine covfie configuration (scaling and rotation) given the size of the field (min and max)
+/// @param nBins 3D array of containing the number of bins for each axis.
+/// @param min (min_x, min_y, min_z)
+/// @param max (max_x, max_y, max_z)
+/// @return The affine field configuration.
+template <typename backend_t, typename point3_1_t, typename point3_2_t,
+          typename point3_3_t>
```
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:107]: The parameters are not of the same type? Not even `min`/`max`?

File: Plugins/Covfie/src/FieldConversion.cpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+/// @brief Creates a covfie field from a magnetic field provider by sampling it.
+/// @param magneticField The acts magnetic field provider.
+/// @param cache The acts cache.
+/// @param nBins 3D array of containing the number of bins for each axis.
+/// @param min (min_x, min_y, min_z)
+/// @param max (max_x, max_y, max_z)
+/// @return A clamp affine linear strided covfie field.
+InterpolatedField covfieField(const Acts::MagneticFieldProvider& magneticField,
```
Comment (paulgessinger) [inline Plugins/Covfie/src/FieldConversion.cpp:177]: Just to not: if you have an `InterpolatedMagneticField` that's held as a `MagneticFieldProvider`, I believe it will go through this overload. Should we test this?

File: Tests/UnitTests/Plugins/Covfie/CovfieFieldConversionTest.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+#include <sstream>
+#include <vector>
+
+// Boost include(s)
+#include <boost/test/unit_test.hpp>
+
+using namespace Acts::UnitLiterals;
+
+template <typename cache_t, typename view_t, typename iterator_t>
+void CheckMagneticFieldEqual(const Acts::MagneticFieldProvider& fieldProvider,
```
Comment (paulgessinger) [inline Tests/UnitTests/Plugins/Covfie/CovfieFieldConversionTest.cpp:31]:
```suggestion
void checkMagneticFieldEqual(const Acts::MagneticFieldProvider& fieldProvider,
```

File: Tests/UnitTests/Plugins/Covfie/CovfieFieldConversionTest.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+#include <cmath>
+#include <sstream>
+#include <vector>
+
+// Boost include(s)
+#include <boost/test/unit_test.hpp>
+
+using namespace Acts::UnitLiterals;
+
+template <typename cache_t, typename view_t, typename iterator_t>
```
Comment (paulgessinger) [inline Tests/UnitTests/Plugins/Covfie/CovfieFieldConversionTest.cpp:30]: You don't need to template on the `cache_t` here.

File: Tests/UnitTests/Plugins/Covfie/CovfieFieldConversionTest.cpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+  for (int i = 0; i < 64; i++) {
+    bField_xyz.push_back(Acts::Vector3(i, i, i));
+  }
+
+  Acts::MagneticFieldContext fieldContext;
+  auto actsField = Acts::fieldMapXYZ(localToGlobalBin_xyz, xPos, yPos, zPos,
+                                     bField_xyz, 1, 1, false);
+  auto cache = actsField.makeCache(fieldContext);
+
+  auto field = Acts::CovfiePlugin::covfieField(actsField);
```
Comment (paulgessinger) [inline Tests/UnitTests/Plugins/Covfie/CovfieFieldConversionTest.cpp:86]: Re general comment:
```suggestion
  Acts::CovfiePlugin::covfieField field(actsField);
```

**Review by paulgessinger — DISMISSED** [review]
General comment: you're **heavily** using `auto`. While it definitely has it's uses, it's usually not a bad idea to type out the types, especially if they're not complicated.
Especially function return types benefit from this, where `auto` is most helpful when the return type is hard to figure out outside of the function body.

None of the comments are blockers, but please consider them anyway.

**Comment by paulgessinger** [comment]
There's also a host of build failures in the CI, do they seem related to you?

File: CMakeLists.txt [hunk 8]
```diff
@@ -308,26 +290,50 @@ endmacro()
...
-    add_subdirectory(thirdparty/covfie)
+
+if (ACTS_BUILD_PLUGIN_TRACCC)
+  if (ACTS_SETUP_VECMEM)
+    if (ACTS_USE_SYSTEM_VECMEM)
+      find_package(vecmem ${_acts_vecmem_version} REQUIRED)
+    else()
+      add_subdirectory(thirdparty/vecmem)
+      # Make the "VecMem language code" available for the whole project.
+      include( "${VECMEM_LANGUAGE_DIR}/vecmem-check-language.cmake" )
```
[the remark this answers is not in the document]
Comment (fredevb) [reply CMakeLists.txt:301]: I think I based the setup on the same setup used in traccc or detray, so I'm not sure if moving it would have side effects.

File: Plugins/Covfie/src/FieldConversion.cpp (hunk 4 above)
Comment (fredevb) [reply Plugins/Covfie/src/FieldConversion.cpp:176]: I have added that the magnetic field provider must be defined within min (inclusive) and max (inclusive) as a comment for the function. For the covfieField function that takes an interpolated acts field, since you don't pass nBins, min, and max, hopefully, it's clear which of the two functions you are using.

File: CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (paulgessinger) [reply CMakeLists.txt:301]: Can you try? If it works fine from inside the subdirectory, that seems cleaner to me.
